### PR TITLE
[MONDRIAN-2398] - Wrong results when aggregate tables used to calcula…

### PR DESCRIPTION
**Pull request by stanislau-strelchanka**
[MONDRIAN-2398] - Wrong results when aggregate tables used to calculate avg value which rolls up one or more NULLs

### CODE REVIEW

**Comment by stanislau-strelchanka** [comment]
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

File: src/main/mondrian/rolap/aggmatcher/ExplicitRecognizer.java [hunk 1]
```diff
@@ -440,6 +447,17 @@ public int compare(
                 }
             });
     }
+
+    protected String getFactCountColumnName
+            (final JdbcSchema.Table.Column.Usage aggUsage) {
+        String measureName = aggUsage.getColumn().getName();
+        Map<String, String> measuresFactCount = tableDef.getMeasuresFactCount();
+        String factCountColumnName = measuresFactCount.get(measureName);
+        if (StringUtils.isEmpty(factCountColumnName)) {
```
Comment (mkambol) [inline src/main/mondrian/rolap/aggmatcher/ExplicitRecognizer.java]: mondrian.olap.Util.isEmpty() is used in most places, probably best to stay consistent.

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                + "    <AggLevel name=\"[Time].[Year]\" column=\"the_year\" />\n"
+                + "    <AggLevel name=\"[Time].[Quarter]\" column=\"quarter\" />\n"
+                + "    <AggLevel name=\"[Time].[Month]\" column=\"month_of_year\" />\n"
+                + "</AggName>\n";
+
+        String schema = getAggSchema(agg);
+
+        verifySameAggAndNot(QUERY, schema);
+    }
+
```
Comment (mkambol) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: I might be missing something, but these tests look identical to each other.
Also, they don't actually validate that an aggtable is used for the avg calculation.  You could verify that expected SQL is fired, instead.

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    protected String getFileName() {
+        return "agg_measure_fact_count_test.csv";
+    }
+
+    @Override
+    public void setUp() throws Exception {
+        propSaver.set(propSaver.properties.GenerateFormattedSql, true);
+        propSaver.set(propSaver.properties.UseAggregates, true);
+        propSaver.set(propSaver.properties.ReadAggregates, true);
+        propSaver.set(propSaver.properties.IgnoreInvalidMembers, true);
```
Comment (mkambol) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: Why IgnoreInvalidMembers=true?

File: src/main/mondrian/olap/Mondrian.xml [hunk 4]
```diff
@@ -1844,6 +1854,19 @@ Revision is $Id$
 
     <Element type="AggFactCount" class= "AggColumnName"/>
 
+    <Element type="AggMeasureFactCount" class= "AggColumnName">
+        <Attribute name="measure" required="false">
```
Comment (stanislau-strelchanka) [inline src/main/mondrian/olap/Mondrian.xml]: @mkambol renamed attribute to 'factColumn'

File: src/main/mondrian/rolap/aggmatcher/ExplicitRecognizer.java (hunk 1 above)
Comment (stanislau-strelchanka) [inline src/main/mondrian/rolap/aggmatcher/ExplicitRecognizer.java]: @mkambol replaced

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java (hunk 3 above)
Comment (stanislau-strelchanka) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: @mkambol removed this line

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java (hunk 2 above)
Comment (stanislau-strelchanka) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: @mkambol fixed tests and added generated sql validation

**Comment by stanislau-strelchanka** [comment]
@mkambol I've added doc description for new element.
Regarding not using agg table if measure fact column is absent, I believe there might be cases 
where only one general fact count column will be enough for all fact table measures and there will be no need in creating multiple measure fact columns. Please correct me if I'm wrong.

**Comment by stanislau-strelchanka** [comment]
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

File: doc/schema.html [hunk 5]
```diff
@@ -3488,6 +3489,10 @@
       	<td><code>&lt;<a name="XML_AggFactCount" href="xml_schema.html#AggFactCount">AggFactCount</a>&gt;</code></td>
       	<td>Specifies name of the column in the candidate aggregate table which contains the number of fact table rows.</td>
     </tr>
+	<tr>
+		<td><code>&lt;<a name="XML_AggMeasureFactCount" href="xml_schema.html#AggMeasureFactCount">AggMeasureFactCount</a>&gt;</code></td>
+		<td>Specifies name of the column in the candidate aggregate table which contains the number of fact table rows associated with a specific measure.</td>
+	</tr>
```
Comment (mkambol) [inline doc/schema.html]: To clarify intent, maybe something like:  "the number of non-null fact table rows associated with a specific fact column.  Mondrian can use this information when calculating averages from aggregate table values."

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java (hunk 2 above)
Comment (mkambol) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: The following tests could still pass even if an agg table wasn't used.

File: testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        testContext.assertQueryReturns(QUERY, result);
+    }
+
+    public void testAggPattern() {
+        String agg = ""
+                + "<AggPattern pattern=\"agg_c_6_fact_csv_2016\">\n"
+                + "    <AggFactCount column=\"FACT_COUNT\"/>\n"
+                + "    <AggMeasureFactCount column=\"store_sales_fact_count\" factColumn=\"store_sales\" />\n"
+                + "    <AggMeasureFactCount column=\"store_cost_fact_count\" factColumn=\"store_cost\" />\n"
+                + "    <AggMeasureFactCount column=\"unit_sales_fact_count\" factColumn=\"unit_sales\" />\n"
```
Comment (lucboudreau) [inline testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java]: Is the factColumn value supposed to be case sensitive? It is different here as compared to the definition within the AggMeasure element. What happens if the DB does not match those? Please try it / test it.

File: src/main/mondrian/olap/Mondrian.xml [hunk 7]
```diff
@@ -1844,6 +1854,19 @@ Revision is $Id$
 
     <Element type="AggFactCount" class= "AggColumnName"/>
 
+    <Element type="AggMeasureFactCount" class= "AggColumnName">
+        <Attribute name="factColumn" required="false">
```
Comment (lucboudreau) [inline src/main/mondrian/olap/Mondrian.xml]: Why is factColumn not required? Are there scenarios where we don't set it? I also don't see any tests where it's not set.

**Comment by stanislau-strelchanka** [comment]
1) made fact count required
2) added test cases for fact count (case sensetive, absent etc)
3) fixed tests to check that aggregate tables are used
4) fixed licences, checkstyle.sh doesn't show any error now
5) updated workbench
6) run test suite with aggregates, but skipped Olap4jTckTest.java - it hangs ( there are errors, but they are the same as in master )
           
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

**Comment by lucboudreau** [comment]
About the Olap4jTckTest tests, we need to fix those. It is not ok to have hanging/failing tests, especially not in the API tests.

**Comment by lucboudreau** [comment]
Everything checks out. The failures are known and due to the Java 8 change. 

@stanislau-strelchanka Would you mind fixing these 4 tests please? We need to grab the new ID of the corresponding segments and update the Segment test.

**Comment by stanislau-strelchanka** [comment]
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son could you please review. 
I've updated ids in segment test.
Didn't touch implementation code, so not sure what caused newly  broken tests.

**Comment by stanislau-strelchanka** [comment]
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

**Comment by lucboudreau** [comment]
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Pentaho Community Edition Project: mondrian ........ SUCCESS [17:52 min]
[INFO] Pentaho Community Edition Project: workbench ....... SUCCESS [ 35.706 s]
[INFO] Pentaho Mondrian Parent POM ........................ SUCCESS [  0.016 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 18:31 min
```

**Pull request merged**